### ISSUE TITLE
Unify toArray() columns filter behavior between Phalcon\Mvc\Model and Phalcon\Mvc\Model\Resultset\[Simple/Complex]

### ISSUE THREAD

**Issue by maurits150** [phalcon/cphalcon#12248]
Hello,

Often when I'm selecting a single model I'm doing e.g. Model::findFirst()->toArray(['id']) to convert it to an array. This works fine and an array with only an id field is returned.

However when I'm selecting multiple records, I kind of expected the behavior to be the same.
Unfortunately when I do Model::find()->toArray(['id']) it returns all columns for reach record, instead of just the id field of each record.

No further errors are thrown. This can be dangerous because you kind of expect it to work the same.

The issue is that a resultset does not support passing column filter in the toArray arguments.

Right now the only solution is to manually loop the resultset and build the array by calling toArray on each record and this is kind of tedious.

It would be great if the behavior could be unified. Thank you.
### Details
- Phalcon version: 3.0.0
- PHP Version: PHP 7.0.9 (cli) (built: Jul 20 2016 11:08:23) ( NTS )
- Operating System: Windows 7
- Server: Nginx

**Comment by Jurigag** [comment]
Just do:

```php
Model::find([
    'columns'=>'id'
])->toArray();
```

This is real solution.

**Comment by Jurigag** [comment]
For me there are not strings. Just change your old mysql driver to mysqlnd and disable stringify in phalcon by adding PDO::ATTR_STRINGIFY_FETCHES => false, to options in db connection, then you will get data as they are in db(like ints, floats, etc)

Well that's obvious that you can't use model methods because it's not full object returned. But why you need them ? Are gonna do something to returned data ? As far as i understand you wanted to have only ids, so you have, you wasn't mentioning first doing something to whole resultset of models.

**Comment by maurits150** [comment]
I think I fixed it by writing my own decorator class for the resultsets, overwriting find and findBy in my models, wrapping the resultsets from that function with my decorator, and inside the decorator made my own toArray function. Which loops through all records and calls toArray with the column filter on each record. [Code](https://maurits.tv/data/file/September%202016/2016-09-25_16-57-46_NMnVSC2Lu9.txt)

Seems to work fine right now. Still think this could be an enhancement to the framework itself though.

**Comment by Jurigag** [comment]
Why just not select proper columns and that's it ? I wrote you above how you can still get ints etc. For relations you have sql joins and basically it.

**Comment by maurits150** [comment]
Well I feel this method is more flexible in the long term. Mainly because my model class is still available on each row. This means that I can still run functions on rows and I can be sure that other model logic also still runs.

Another issue I faced with the default Resultset::toArray() is that afterFetch doesn't run on each row, which means that some of my columns don't get decoded properly (I have a some database columns that hold JSON from external APIs which are encoded/decoded on beforeSave and afterFetch.)

Writing my own decorator that overrides Resultset::toArray() and manually toArray's each row solved this issue too because it actually creates each model (instead of returning raw query data)

**Comment by Jurigag** [comment]
Yea, but it's much slower. `Resultset::toArray()` is really super duper fast, mostly because `afterFetch` is not fired, even not selecting columns and doing `toArray()` will not fire afterFetch, it just return whole resultset as an array. `aftetFetch` is fired only when you access full model in loop.

**Comment by maurits150** [comment]
Yeah you're right on that. I'm not selecting large amounts of data luckily so it's not a problem.

IMO having consistent behavior across the orm is more important then writing faster toArray functions that end up not running things such as afterFetch, especially when you would kind of expect it would because in all other cases it does seem to work fine.

For speed I would consciously choose to use raw query and a fetchAll. or at least make it an optional argument so we're aware that the returned data is unprocessed.

So when I'm using any function inside \Phalcon\Mvc\Model* or in underlying namespaces/classes I don't expect to get unprocessed data directly from \Phalcon\Db.

Basically the result of any find operation should go through the appropriate model logic. (the only exception of course being the retrieval of incomplete models such as with 'columns' option)

**Comment by dschissler** [comment]
I've had a recurring thought that the `afterFetch` function should be something like an anonymous function or a static method on the class.  So something more like a mapping function from one array to another and having no context.  By having `afterFetch` be a normal event it requires first converting the data into an Model object and then possibly reconverting from there.  So the `Resultset::toArray()` technique just circumvents this bad design and would work in many cases unless complex data like JSON was used.
